### DISCUSSION TITLE
how to get faster reconnects from deep sleep

### DISCUSSION THREAD

**Discussion by toaster-robotics** [h2zero/NimBLE-Arduino#680] [General]
Hello,
I currently have a project that uses a ESP32-s3 as a client. It connects with a BLE remote(server) that acts as HID keyboard. I just got everything working and it's catching all the button presses. However, the remote goes into deep sleep after 10 minutes of inactivity. Any button press wakes it up and the ESP32 automatically sees it and reconnects. The issues is that it takes ~2 seconds from starting to completing the reconnection. So any button presses in between those 2 seconds are not caught. This makes it a little awkward to use. Escpeclliy for the non-tech people I am targeting for use. I have a BLE mouse that also goes into deep sleep after a period of inactivity. But it reconnects on my laptop instantly. It seems like sub 1 second. So I know it's possible to reconnect faster. I did a few hours of internet sleuthing and wasn't able to find anything that was helpful. I did come across this [post](https://www.esp32.com/viewtopic.php?t=36025) that references the [blemouse2xac](https://github.com/touchgadget/blemouse2xac) project. However I didn't find anything there that I'm not already doing. Any ideas on how to speed up the reconnections? Thanks

![23d8513e-2bd6-4827-92d6-48e13f59eeea __CR0,0,970,600_PT0_SX970_V1___](https://github.com/h2zero/NimBLE-Arduino/assets/141705577/98175349-c927-48a1-8ee3-439e54bf3a98)

**Comment by toaster-robotics** [comment]
I was able to get the reconnect down to ~0.75 seconds by specifying the **NimBLERemoteCharacteristic** to subscribe to. I have it pared down to 3 **NimBLERemoteCharacteristic** or 3 handles (23, 31, 35). The bulk of the reconnect time is subscribing to the characteristics. One (23) takes ~0.05 seconds to subscribe. But the next two (31, 35) take ~0.35 seconds each. I couldn't get it to go any faster.
One strange thing is that I only need two (23, 31) of the characteristics to catch button presses. But for some reason, if I don't subscribe to the third one (35) I never receive notifications on the other two. If I can get this part figured out then I can cut the time down to ~0.4 seconds.

**Comment by tapir** [comment]
Is there a sample code you can share? I'm exactly in the same boat except light sleep instead of deep. Also may I ask how did you manage to wake up from deep sleep with so many keys? AFAIK waking up from deep sleep is only possible with RTC pins which are very limited.

**Comment by toaster-robotics** [comment]
@tapir , I just made my code into a [library](https://github.com/toaster-robotics/bleM3Remote). I can't speak for your device. But the remote I'm using wakes with any button press. I have done 0 modifications to the remote. I've only been working on the ESP32 as a client. I don't touch the server. The remote goes for close to $1 on Aliexpress if you want to check it out.

I'm kinda thinking to achieve near instantaneous reconnects from sleep (like on a firestick remote) there would need to be changes to the BLE remote code. But I don't plan on attempting that because: 1. I'm treating it as a off-the-shelf part and 2. I would have no idea how to do that.

**Comment by richardrob24** [comment]
I'm having the same issue.

Based on what I've read from the [BT HID Spec](https://www.bluetooth.com/specifications/specs/html/?src=HID_v1.1.1/out/en/index-en.html#UUID-57148acd-9a3a-c538-5d37-4a6237b96e07), we need to add the HIDReconnectInitiate and HIDVirtualCable attributes to the input service. In order for that to work, you have to store some of the client's information in non-volatile memory on your device. Once you have that, when you wake up you can just retrieve the client's parameters and immediately send a keystroke report without having to go through the initial connection setup process. 

I'm 99% sure neither NimBLE nor Bludroid support these yet, so I'll have to work on implementing this feature. I have no idea how long this will take me.
